openmmml/models/macepotential.py

```diff
@@ -0,0 +1,222 @@
+USE OR OTHER DEALINGS IN THE SOFTWARE.
+"""
+
+from openmmml.mlpotential import MLPotential, MLPotentialImpl, MLPotentialImplFactory
+import openmm
+from typing import Iterable, Optional, Union, Tuple
+from openmmml.models.utils import simple_nl
+import logging
+
+
+class MACEPotentialImplFactory(MLPotentialImplFactory):
+    """This is the factory that creates MACEPotentialImpl objects."""
+
+    def createImpl(self, name: str, model_path: str, **args) -> MLPotentialImpl:
+        return MACEPotentialImpl(name, model_path)
+
+class MACEPotentialImpl(MLPotentialImpl):
+    """This is the MLPotentialImpl implementing the MACE potential.
+
+    The potential is implemented using MACE to build a PyTorch model.  A
+    TorchForce is used to add it to the OpenMM System.
+
+    You must specify the path to a trained MACE model e.g.
+    >>> potential = MLPotential('mace', model_path='MACE.model')
+
+    When you create the system can optionally change the dtype of the model
+    with the keyword argument dtype='float32' or dtype='float64'. The default behavoir is
+    to use the dtype of the loaded MACE model.
+    e.g.
+    >>> system = potential.createSystem(topology, dtype='float32')
+
+    You can also specify that the full atomic energy (including the atom self energy) is returned 
+    rather than the default of the interaction energy.
+    e.g.
+    >>> system = potential.createSystem(topology, interaction_energy=False)
+    
+    """
+
+    def __init__(self, name, model_path):
+        self.name = name
+        self.model_path = model_path
+
+    def addForces(self,
+                  topology: openmm.app.Topology,
+                  system: openmm.System,
+                  atoms: Optional[Iterable[int]],
+                  forceGroup: int,
+                  implementation : str=None,
+                  dtype: str=None,
+                  interaction_energy: bool=True,
+                  **args):
+        
+
+        import torch
+        import openmmtorch
+        from e3nn.util import jit
+        from mace.tools import utils, to_one_hot, atomic_numbers_to_indices
+        
+        includedAtoms = list(topology.atoms())
+        if atoms is not None:
+            includedAtoms = [includedAtoms[i] for i in atoms]
+        
+        class MACEForce(torch.nn.Module):
+
+            def __init__(self, model_path, atomic_numbers, indices, periodic, dtype=None, interaction_energy=True):
+                super(MACEForce, self).__init__()
+
+
+                if not torch.cuda.is_available():
+                    self.model = torch.load(model_path, map_location=torch.device('cpu'))
+                else:
+                    self.model = torch.load(model_path)
+                
+                # get the dtype of the saved model
+                model_dtype = [p.dtype for p in self.model.parameters()][0]
+
+                # check is user has request a different dtype    
```

Comment:
The model should run on the precision with which it was trained. If OpenMM runs at a different precision (double rather float), one just casts inputs/outputs accordingly rather than change the model.